MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -212,208 +217,343 @@ double GeoMapper::getMeshElevation(
 }
 
-bool GeoMapper::isPntInBoundingBox(double ax, double ay, double bx, double by, double px, double py) const
+static void insertSubSegments(
+    GeoLib::Polyline& ply, GeoLib::PointVec& points,
+    GeoLib::Polyline::SegmentIterator& segment_it,
+    std::vector<GeoLib::LineSegment> const& sub_segments)
 {
-    if ( px < (std::min(ax, bx)-std::numeric_limits<double>::epsilon()) || px > (std::max(ax, bx)+std::numeric_limits<double>::epsilon()) ||
-         py < (std::min(ay, by)-std::numeric_limits<double>::epsilon()) || py > (std::max(ay, by)+std::numeric_limits<double>::epsilon()) )
-         return false;
-    return true;
+    std::size_t const j(segment_it.getSegmentNumber());
+    std::size_t new_pnts_cnt(0);
+    for (auto const& segment : sub_segments)
+    {
+        auto const begin_id(points.push_back(
+            new GeoLib::Point(segment.getBeginPoint(), points.size())));
+        if (ply.insertPoint(j + new_pnts_cnt + 1, begin_id))
+            new_pnts_cnt++;
+        auto const end_id(points.push_back(
+            new GeoLib::Point(segment.getEndPoint(), points.size())));
+        if (ply.insertPoint(j + new_pnts_cnt + 1, end_id))
+            new_pnts_cnt++;
+    }
+    segment_it += new_pnts_cnt;
 }
 
-double GeoMapper::getMaxSegmentLength(const std::vector<GeoLib::Polyline*> &lines) const
+static void mapPolylineOnSurfaceMesh(
+    GeoLib::Polyline& ply,
+    GeoLib::PointVec& orig_points,
+    MeshLib::MeshElementGrid const& mesh_element_grid)
 {
-    double max_segment_length (0);
-    const std::size_t nPlys ( lines.size() );
-    for (std::size_t i=0; i<nPlys; ++i)
+    // for each segment ...
+    for (auto segment_it(ply.begin()); segment_it != ply.end(); ++segment_it)
     {
-        const GeoLib::Polyline* line = lines[i];
-        const std::size_t nPlyPoints = line->getNumberOfPoints();
-        for (std::size_t j=1; j<nPlyPoints; ++j)
+        auto candidate_elements(getCandidateElementsForLineSegmentIntersection(
+            mesh_element_grid, *segment_it));
+
+        auto mapPoint = [&candidate_elements](MathLib::Point3d& p) {
+            auto const* elem(
+                findElementContainingPoint2d(candidate_elements, p));
+            if (elem)
+            {
+                if (!snapPointToElementNode(p, *elem, 1e-3))
+                    mapPointOnSurfaceElement(*elem, p);
+            }
+            return elem;
+        };
+
+        // map segment begin and end point
+        auto const* beg_elem(mapPoint((*segment_it).getBeginPoint()));
+        auto const* end_elem(mapPoint((*segment_it).getEndPoint()));
+
+        // Since the mapping of the segment begin and end points the coordinates
+        // changed. The internal data structures of PointVec are possibly
+        // invalid and hence it is necessary to re-create them.
+        orig_points.resetInternalDataStructures();
+
+        if (beg_elem == end_elem)  // TODO: handle case: beg_elem == end_elem == nullptr
+            continue;
```

Comment:
I assume, that the mentioned todo case does not happen (yet); Then it would be helpful if such a case happens, to have an assert inside the if.
```c++
if (b == e) {assert(b != nullptr}; continue;}
```